[photon-lib] Python support for PNP_DISTANCE_TRIG_SOLVE

kcooney opened the pull request:
This adds support for PNP_DISTANCE_TRIG_SOLVE in the the python PhotonPoseEstimator, mirroring the implementation in the Java PhotonPoseEstimator.

Changes:
- Add PoseStrategy.PNP_DISTANCE_TRIG_SOLVE
- Add PhotonPoseEstimator.addHeadingData()
- Update PhotonCameraSim to optionally take in a function to get the FPGA timestamp
- Fix PhotonCameraSim.process() to set ntReceiveTimestampMicros in the result
- Minor readability improvements to PhotonPipelineResult
- Minor test improvements to PhotonPoseEstimatorTest
- Add .vscode/settings.json (to make running pythonb tests in VSCode easier)

Merge checklist:
- [x] Pull Request title is [short, imperative summary](https://cbea.ms/git-commit/) of proposed changes
- [x] The description documents the _what_ and _why_
- [ ] If this PR changes behavior or adds a feature, user documentation is updated
- [ ] If this PR touches photon-serde, all messages have been regenerated and hashes have not changed unexpectedly
- [ ] If this PR touches configuration, this is backwards compatible with settings back to v2024.3.1
- [ ] If this PR touches pipeline settings or anything related to data exchange, the frontend typing is updated
- [ ] If this PR addresses a bug, a regression test for it is added

samfreund commented:
Looks like you need to lint, if you're not sure how to do that I can also do it for you.

Gold856 reviewed (commented):
Looks mostly good to me. I'm curious as to why the timestamp injection is needed though.

kcooney commented:
> So, is it just too fast?

@Gold856 I don't know how the HAL timers are started in the Java code, so I can't answer that question with much certainty.

> If it's too fast, I'll accept a small sleep at the start of the test with a comment explaining why to make it work. I really don't like having that timestamp function.

I strongly dislike adding sleeps to "fix" tests. 

I changed the tests to set `PhotonPipelineResult.ntReceiveTimestampMicros` to avoid the negative timestamps, which allowed me to remove the `timestampFunc` parameter that I had added.

> Alternatively, it looks like we use PhotonCameraInjector in order to inject pipeline results, which I suspect is related, since it allows you to specify the timestamp yourself. You could give that a try as well.

The code that is calling `getFPGATimestamp()` is in `PhotonCameraSim`, not `PhotonCamera`, so I can't fix this in `PhotonCameraInjector`.

I did a bit of searching, and it looks like I could have the test use `wpilib.simulation.setRuntimeType()` to have the test use simulated time. I couldn't find any tests in PhotonVision or WPILib that do that, so I didn't go with that approach.

mcm001 commented on photon-lib/src/test/java/org/photonvision/PhotonPoseEstimatorTest.java:
```diff
@@ -526,63 +527,63 @@ void closestToLastPose() {
-        assertEquals(realPose.getY(), pose.getY(), .01);
-        assertEquals(0.0, pose.getZ(), .01);
+        try (PhotonCameraSim cameraOneSim =
+                new PhotonCameraSim(cameraOne, SimCameraProperties.PERFECT_90DEG())) {
```
Why did Java change?

kcooney replied:
@mcm001  Two changes here
- I saw a warning because `PhotonCameraSim` implements `Closeable` but we didn't close `cameraOneSim` here
- I added an additional assertion to verify that the position was based on the correct April tag (initially the python tests were failing because the estimator produced a different pose, and I added the assertion to both to narrow down where the bug was)

I'd be happy to revert the first change, and move that to another PR (possibly fixing that warning everywhere).

mcm001 replied:
nah we ball
